### PR TITLE
[FIX] im_livechat: fix the session duration

### PR DESCRIPTION
**Pull request by std-odoo**
Bug
===
The session duration added in the KPI have the wrong
value because of a typo in the SQL query.

Task-2827996

### CODE REVIEW

**Review by tde-banana-odoo — APPROVED** [review]

**Comment by tde-banana-odoo** [comment]
Note that upgrade is pas content for some reason ...

**Comment by std-odoo** [comment]
@tde-banana-odoo I think it's not happy because it change the postgre SQL view, but I don't think it can cause issue in stable (as this is just a view and not "real SQL rows"

File: addons/im_livechat/models/digest.py [hunk 1]
```diff
@@ -12,7 +12,7 @@ class Digest(models.Model):
     kpi_livechat_conversations = fields.Boolean('Conversations handled')
     kpi_livechat_conversations_value = fields.Integer(compute='_compute_kpi_livechat_conversations_value')
     kpi_livechat_response = fields.Boolean('Time to answer(sec)', help="Time to answer the user in second.")
-    kpi_livechat_response_value = fields.Float(compute='_compute_kpi_livechat_response_value')
+    kpi_livechat_response_value = fields.Float(digits=(16, 2), compute='_compute_kpi_livechat_response_value')
```
Comment (awa-odoo) [inline addons/im_livechat/models/digest.py:15]: Does this change the DB schema (definition of the column) ?

File: addons/im_livechat/report/im_livechat_report_channel.py [hunk 2]
```diff
@@ -90,9 +90,9 @@ def init(self):
                     END as is_unrated,
                     C.livechat_operator_id as partner_id
                 FROM mail_channel C
-                    JOIN mail_message M ON (M.res_id = C.id AND m.model = 'mail.channel')
+                    JOIN mail_message M ON (M.res_id = C.id AND M.model = 'mail.channel')
```
Comment (awa-odoo) [inline addons/im_livechat/report/im_livechat_report_channel.py:93]: Wow...

2 questions regarding this:

- Is it going to be applied for clients in stable? Don't quite remember when "init" is called. But if it's only at module installation/update then isn't it too late anyway?
- Is there some way to make an automated test for this?

**Review by awa-odoo — COMMENTED** [review]
Hello @std-odoo 

Nice find!
I left some questions to make sure I understand how it will impact stable.

Thanks & cheers!

**Review by std-odoo — COMMENTED** [review]
@awa-odoo Hi :) 

I did the changes and answered :)

**Review by awa-odoo — COMMENTED** [review]
Hello @std-odoo 

Thanks for taking the time to write a test.
Added a few comments.

Cheers!

File: addons/im_livechat/tests/test_im_livechat_report.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        self.operator = channel.livechat_operator_id
+
+        message_1 = channel.message_post(author_id=self.visitor_user.partner_id.id, body="Hello")
+        message_2 = channel.message_post(author_id=self.operator.id, body="Hi")
+        message_3 = channel.message_post(author_id=self.operator.id, body="Hi")
+        self._utils_set_create_date(message_1, "2023-03-17 06:05:54.524067")
+        self._utils_set_create_date(message_2, "2023-03-17 08:15:54.524067")
+        self._utils_set_create_date(message_3, "2023-03-17 08:45:54.524067")
+
+        fake_message_1 = channel.message_post(author_id=self.operator.id, body="1")
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/im_livechat/tests/test_im_livechat_report.py]: yes, fake was a bad name, it just to ensure we filter on the right model (what the old code doesn't do)

File: addons/im_livechat/tests/test_im_livechat_report.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        fake_message_2 = channel.message_post(author_id=self.operator.id, body="2")
+        self._utils_set_create_date(fake_message_1, "2023-03-17 05:05:54.524067")
+        self._utils_set_create_date(fake_message_2, "2023-03-17 09:15:54.524067")
+        (fake_message_1 | fake_message_2).model = "res.partner"
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/im_livechat/tests/test_im_livechat_report.py]: yes exactly :) I added a comment

File: addons/im_livechat/tests/test_im_livechat_report.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        self.env['mail.message'].flush()
+
+    def test_im_livechat_report_channel(self):
+        report = self.env['im_livechat.report.channel'].search([('livechat_channel_id', '=', self.livechat_channel.id)])
+        self.assertEqual(len(report), 1, "Should have one channel report per channel")
+        self.assertEqual(int(report.time_to_answer), 7800)
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/im_livechat/tests/test_im_livechat_report.py:43]: duration is in seconds, ok I added a comment

File: addons/im_livechat/tests/test_im_livechat_report.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        self.assertEqual(int(report.duration), 9600)
+
+    def test_im_livechat_report_operator(self):
+        result = self.env['im_livechat.report.operator'].read_group([], ['time_to_answer:avg', 'duration:avg'], [])
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/im_livechat/tests/test_im_livechat_report.py:47]: yes, to variate a bit (and make sur the read group work well on the SQL view, etc)

File: addons/im_livechat/tests/test_im_livechat_report.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        self.assertEqual(len(result), 1)
+        self.assertEqual(int(result[0]['time_to_answer']), 7800)
+        self.assertEqual(int(result[0]['duration']), 9600)
+
+    def _utils_set_create_date(self, records, date_str):
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/im_livechat/tests/test_im_livechat_report.py]: with freeze time it's not possible, but I can mock `self.env.cr.now` and it work :)

**Review by std-odoo — COMMENTED** [review]
@awa-odoo Hi, changes done :)

**Pull request closed without merging**